File: backend/models/Property.js
const mongoose = require('mongoose');

const propertySchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'Please provide a property title'],
    trim: true,
    maxlength: [100, 'Title cannot be more than 100 characters']
  },
  description: {
    type: String,
    required: [true, 'Please provide a description'],
    maxlength: [2000, 'Description cannot be more than 2000 characters']
  },
  price: {
    type: Number,
    required: [true, 'Please provide a price']
  },
  propertyType: {
    type: String,
    required: [true, 'Please specify property type'],
    enum: ['Apartment', 'House', 'Villa', 'Land', 'Office', 'Studio']
  },
  status: {
    type: String,
    required: [true, 'Please specify status'],
    enum: ['For Sale', 'For Rent']
  },
  area: {
    type: Number, // in square meters
    required: [true, 'Please provide area']
  },
  bedrooms: {
    type: Number,
    default: 0
  },
  bathrooms: {
    type: Number,
    default: 0
  },
  averageRating: {
    type: Number,
    default: 0
  },
  reviewCount: {
    type: Number,
    default: 0
  },
  location: {
    address: {
      type: String,
      required: [true, 'Please provide an address']
    },
    city: {
      type: String,
      required: [true, 'Please provide a city']
    },
    state: String,
    country: {
      type: String,
      default: 'Morocco'
    },
    zipCode: String,
    coordinates: {
      latitude: {
        type: Number,
        required: true
      },
      longitude: {
        type: Number,
        required: true
      }
    }
  },
  images: [{
    type: String
  }],
  amenities: [{
    type: String
  }],
  owner: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  isAvailable: {
    type: Boolean,
    default: true
  },
  views: {
    type: Number,
    default: 0
  },
  featured: {
    type: Boolean,
    default: false
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
});

// Update the updatedAt timestamp before saving
propertySchema.pre('save', function(next) {
  this.updatedAt = Date.now();
  next();
});

// Index for searching
propertySchema.index({ title: 'text', description: 'text' });
propertySchema.index({ 'location.city': 1, price: 1, propertyType: 1 });

const Property = mongoose.model('Property', propertySchema);

module.exports = Property;

File: mobile_app/lib/models/property.dart
import 'user.dart';

class Property {
  final String id;
  final String title;
  final String description;
  final double price;
  final String propertyType;
  final String status;
  final double area;
  final int bedrooms;
  final int bathrooms;
  final PropertyLocation location;
  final List<String> images;
  final List<String> amenities;
  final User? owner;
  final bool isAvailable;
  final int views;
  final bool featured;
  final double averageRating;
  final int reviewCount;
  final DateTime createdAt;
  final DateTime updatedAt;

  Property({
    required this.id,
    required this.title,
    required this.description,
    required this.price,
    required this.propertyType,
    required this.status,
    required this.area,
    this.bedrooms = 0,
    this.bathrooms = 0,
    required this.location,
    this.images = const [],
    this.amenities = const [],
    this.owner,
    this.isAvailable = true,
    this.views = 0,
    this.featured = false,
    this.averageRating = 0.0,
    this.reviewCount = 0,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Property.fromJson(Map<String, dynamic> json) {
    return Property(
      id: json['_id'] ?? json['id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      price: (json['price'] ?? 0).toDouble(),
      propertyType: json['propertyType'] ?? '',
      status: json['status'] ?? '',
      area: (json['area'] ?? 0).toDouble(),
      bedrooms: json['bedrooms'] ?? 0,
      bathrooms: json['bathrooms'] ?? 0,
      location: PropertyLocation.fromJson(json['location'] ?? {}),
      images: json['images'] != null
          ? List<String>.from(json['images'])
          : [],
      amenities: json['amenities'] != null
          ? List<String>.from(json['amenities'])
          : [],
      owner: json['owner'] != null ? User.fromJson(json['owner']) : null,
      isAvailable: json['isAvailable'] ?? true,
      views: json['views'] ?? 0,
      featured: json['featured'] ?? false,
      averageRating: (json['averageRating'] ?? 0).toDouble(),
      reviewCount: json['reviewCount'] ?? 0,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'title': title,
      'description': description,
      'price': price,
      'propertyType': propertyType,
      'status': status,
      'area': area,
      'bedrooms': bedrooms,
      'bathrooms': bathrooms,
      'location': location.toJson(),
      'images': images,
      'amenities': amenities,
      'owner': owner?.toJson(),
      'isAvailable': isAvailable,
      'views': views,
      'featured': featured,
      'averageRating': averageRating,
      'reviewCount': reviewCount,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

class PropertyLocation {
  final String address;
  final String city;
  final String? state;
  final String country;
  final String? zipCode;
  final Coordinates coordinates;

  PropertyLocation({
    required this.address,
    required this.city,
    this.state,
    this.country = 'Morocco',
    this.zipCode,
    required this.coordinates,
  });

  factory PropertyLocation.fromJson(Map<String, dynamic> json) {
    return PropertyLocation(
      address: json['address'] ?? '',
      city: json['city'] ?? '',
      state: json['state'],
      country: json['country'] ?? 'Morocco',
      zipCode: json['zipCode'],
      coordinates: Coordinates.fromJson(json['coordinates'] ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'address': address,
      'city': city,
      'state': state,
      'country': country,
      'zipCode': zipCode,
      'coordinates': coordinates.toJson(),
    };
  }
}

class Coordinates {
  final double latitude;
  final double longitude;

  Coordinates({
    required this.latitude,
    required this.longitude,
  });

  factory Coordinates.fromJson(Map<String, dynamic> json) {
    return Coordinates(
      latitude: (json['latitude'] ?? 0).toDouble(),
      longitude: (json['longitude'] ?? 0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'latitude': latitude,
      'longitude': longitude,
    };
  }
}
